### PR TITLE
added new interfaces for RF GPU kernels

### PR DESCRIPTION
**Pull request by makart19**
added new interfaces for RF GPU kernels

### CODE REVIEW

File: cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_classification_hist.cpp [hunk 1]
```diff
@@ -23,7 +24,7 @@ struct train_kernel_cpu<Float, Task, method::hist> {
     train_result<Task> operator()(const dal::backend::context_cpu& ctx,
                                   const descriptor_base<Task>& params,
                                   const train_input<Task>& input) const {
-        return train_result<Task>();
+        throw unimplemented_error("Decision forest train hist method is not implemented for CPU!");
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_classification_hist.cpp:27]:
```suggestion
        throw unimplemented_error("Decision forest train hist method is not implemented for CPU");
```

File: cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_regression_hist.cpp [hunk 2]
```diff
@@ -23,7 +24,7 @@ struct train_kernel_cpu<Float, Task, method::hist> {
     train_result<Task> operator()(const dal::backend::context_cpu& ctx,
                                   const descriptor_base<Task>& params,
                                   const train_input<Task>& input) const {
-        return train_result<Task>();
+        throw unimplemented_error("Decision forest train hist method is not implemented for CPU!");
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_regression_hist.cpp:27]:
```suggestion
        throw unimplemented_error("Decision forest train hist method is not implemented for CPU");
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel_dense_dpc.cpp [hunk 3]
```diff
@@ -23,7 +24,7 @@ struct infer_kernel_gpu<Float, Task, method::dense> {
     infer_result<Task> operator()(const dal::backend::context_gpu& ctx,
                                   const descriptor_base<Task>& params,
                                   const infer_input<Task>& input) const {
-        return infer_result<Task>();
+        throw unimplemented_error("Decision forest infer dense method is not implemented for GPU!");
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel_dense_dpc.cpp:27]:
```suggestion
        throw unimplemented_error("Decision forest infer dense method is not implemented for GPU");
```

File: cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_regression_hist.cpp [hunk 4]
```diff
@@ -15,6 +15,7 @@
 *******************************************************************************/
 
 #include "oneapi/dal/algo/decision_forest/backend/cpu/train_kernel.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_regression_hist.cpp:18]:
```suggestion
#include "oneapi/dal/exceptions.hpp"
```

File: cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_classification_hist.cpp [hunk 5]
```diff
@@ -15,6 +15,7 @@
 *******************************************************************************/
 
 #include "oneapi/dal/algo/decision_forest/backend/cpu/train_kernel.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/cpu/train_kernel_classification_hist.cpp:18]:
```suggestion
#include "oneapi/dal/exceptions.hpp"
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel_dense_dpc.cpp [hunk 6]
```diff
@@ -15,6 +15,7 @@
 *******************************************************************************/
 
 #include "oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/infer_kernel_dense_dpc.cpp:18]:
```suggestion
#include "oneapi/dal/exceptions.hpp"
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#define DAAL_SYCL_INTERFACE
+#define DAAL_SYCL_INTERFACE_USM
+#define DAAL_SYCL_INTERFACE_REVERSED_RANGE
+
+#include <daal/include/services/error_handling.h>
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp:21]: Do you really need this include?

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_dense_dpc.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@
 *******************************************************************************/
 
 #include "oneapi/dal/algo/decision_forest/backend/gpu/train_kernel.hpp"
+#include "oneapi/dal/backend/interop/error_converter.hpp"
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_dense_dpc.cpp:18]:
```suggestion
#include "oneapi/dal/exceptions.hpp"
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_dense_dpc.cpp [hunk 9]
```diff
@@ -23,7 +24,7 @@ struct train_kernel_gpu<Float, Task, method::dense> {
     train_result<Task> operator()(const dal::backend::context_gpu& ctx,
                                   const descriptor_base<Task>& params,
                                   const train_input<Task>& input) const {
-        return train_result<Task>();
+        throw unimplemented_error("Decision forest train dense method is not implemented for GPU!");
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_dense_dpc.cpp:27]:
```suggestion
        throw unimplemented_error("Decision forest train dense method is not implemented for GPU");
```

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#define DAAL_SYCL_INTERFACE
+#define DAAL_SYCL_INTERFACE_USM
+#define DAAL_SYCL_INTERFACE_REVERSED_RANGE
+
+#include <daal/include/services/error_handling.h>
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:21]: Do you really need this include?

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    train_result<Task> res;
+
+    auto daal_result = reg::training::Result();
+
+    /* init daal result's objects */
+    array<Float> arr_oob_err;
+    if (desc.get_train_results_to_compute() &
+        static_cast<std::uint64_t>(train_result_to_compute::compute_out_of_bag_error)) {
+        arr_oob_err = array<Float>::empty(1 * 1);
+        //res.set_oob_err(homogen_table_builder{}.reset(arr_oob_err, 1, 1).build());
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:109]: remove

**Comment by makart19** [comment]
ci http://intel-ci.intel.com/eac1293f-1063-f186-bbb1-8cdcd4b723a1

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+        const auto res_oob_err =
+            interop::convert_to_daal_sycl_homogen_table(queue, arr_oob_err, 1, 1);
+        daal_result.set(reg::training::outOfBagError, res_oob_err);
+    }
+
+    array<Float> arr_oob_per_obs_err;
+    if (desc.get_train_results_to_compute() &
+        static_cast<std::uint64_t>(
+            train_result_to_compute::compute_out_of_bag_error_per_observation)) {
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:119]:
```suggestion
            train_result_to_compute::compute_out_of_bag_error_per_observation) != 0) {
```

Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:119]: the same for other

**Comment by mchernov-intel** [comment]
@makart19 please, provide also precondition, postcondition and invariant checks

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp (hunk 10 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:21]: Yes, i need #define DAAL_SYCL_INTERFACE_REVERSED_RANGE to have correct order in ranges assignment

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp (hunk 12 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:119]: It isn't required here. More over it doesn't make it more informative.

**Comment by makart19** [comment]
>>> precondition, postcondition and invariant checks
Yes, i am going to add it after having this request merged.

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+#include <daal/include/algorithms/decision_forest/decision_forest_classification_training_batch.h>
+#include <daal/include/algorithms/decision_forest/decision_forest_classification_training_types.h>
+
+#include <daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_train_hist_kernel_oneapi.h>
+
+#include "oneapi/dal/algo/decision_forest/backend/gpu/train_kernel.hpp"
+#include "oneapi/dal/algo/decision_forest/backend/interop_helpers.hpp"
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp:33]: common_dpc.hpp includes common.hpp I believe

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+
+#include <daal/include/algorithms/decision_forest/decision_forest_regression_training_batch.h>
+#include <daal/include/algorithms/decision_forest/decision_forest_regression_training_types.h>
+
+#include <daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_kernel_oneapi.h>
+
+#include "oneapi/dal/algo/decision_forest/backend/gpu/train_kernel.hpp"
+#include "oneapi/dal/algo/decision_forest/backend/interop_helpers.hpp"
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/backend/interop/common_dpc.hpp"
```
Comment (amgrigoriev) [inline cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:33]: The same comment

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp (hunk 11 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp]: thanks, removed

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp (hunk 13 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_classification_hist_dpc.cpp:32]: thanks, fixed

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp (hunk 14 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp:32]: thanks, fixed

File: cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp (hunk 10 above)
Comment (makart19) [reply cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_regression_hist_dpc.cpp]: regarding include -> removed it, thanks

**Comment by makart19** [comment]
latets ci http://intel-ci.intel.com/eac1407f-dfea-f156-a46f-8cdcd4b723a1

**Pull request merged**